src/OrleansAzureUtils/Providers/Streams/AzureQueue/AzureQueueBatchContainer2.cs

```diff
@@ -0,0 +1,95 @@
+        /// </summary>
+        /// <param name="original">The object to create a copy of</param>
+        /// <returns>The copy.</returns>
+        public static object DeepCopy(object original)
+        {
+            var source = original as AzureQueueBatchContainer2;
+            if (source == null)
+            {
+                throw new ArgumentNullException(nameof(original));
+            }
+
+            var token = source.sequenceToken == null ? null : new EventSequenceToken2(source.sequenceToken.SequenceNumber, source.sequenceToken.EventIndex);
+            var events = source.events?.Select(SerializationManager.DeepCopy).ToList();
+            var context = source.requestContext?.ToDictionary(kv => kv.Key, kv => SerializationManager.DeepCopy(kv.Value));
+            var copy = new AzureQueueBatchContainer2(source.StreamGuid, source.StreamNamespace, events, context, token);
+            SerializationContext.Current.RecordObject(original, copy);
+            return copy;
+        }
+
+        /// <summary>
+        /// Serializes the container to the binary stream.
+        /// </summary>
+        /// <param name="untypedInput">The object to serialize</param>
+        /// <param name="writer">The stream to write to</param>
+        /// <param name="expected">The expected type</param>
+        public static void Serialize(object untypedInput, BinaryTokenStreamWriter writer, Type expected)
+        {
+            var typed = untypedInput as AzureQueueBatchContainer2;
+            if (typed == null)
+            {
+                throw new SerializationException();
+            }
+
+            writer.Write(typed.StreamGuid);
+            writer.Write(typed.StreamNamespace);
+            SerializationManager.SerializeInner(typed.sequenceToken, writer, typed.sequenceToken?.GetType());
+            SerializationManager.SerializeInner(typed.events, writer, typed.events?.GetType());
+            SerializationManager.SerializeInner(typed.requestContext, writer, typed.requestContext?.GetType());
+        }
+
+        /// <summary>
+        /// Deserializes the container from the data stream.
+        /// </summary>
+        /// <param name="expected">The expected type</param>
+        /// <param name="reader">The stream reader</param>
+        /// <returns>The deserialized value</returns>
+        public static object Deserialize(Type expected, BinaryTokenStreamReader reader)
+        {
+            var guid = reader.ReadGuid();
+            var ns = reader.ReadString();
+            var eventToken = SerializationManager.DeserializeInner<EventSequenceToken2>(reader);
+            var events = SerializationManager.DeserializeInner<List<object>>(reader);
+            var context = SerializationManager.DeserializeInner<Dictionary<string, object>>(reader);
+            var deserialized = new AzureQueueBatchContainer2(guid, ns, events, context, eventToken);
+            DeserializationContext.Current.RecordObject(deserialized);
```

Comment:
@ReubenBond I'll make the change. Although it's unlikely that a cycle will be introduced, it's possible.